File: Бойко Дмитро/Лабораторна робота №1/Customer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab_1
{
    class Customer
    {
        public int ID { get; set; }
        public string Name { get; set; }
        public int Age { get; set; }
        public Operator[] Operators { get; set; }
        public Bill[] Bills { get; set; }
        public double LimitingAmount { get; set; }

        public Customer(int iD, string name, int age, Operator[] operators, Bill[] bills, double limitingAmount)
        {
            ID = iD;
            Name = name;
            Age = age;
            Operators = operators;
            Bills = bills;
            LimitingAmount = limitingAmount;
        }
        public void Talk(int minutes, Customer other)
        {
            if (Operators != null && Operators.Length > 0)
            {
                double cost = Operators[0].CalculateTalkingCost(minutes, this);
                if (Age < 18 || Age > 65)
                {
                    cost *= (1 - Operators[0].DiscountRate / 100.0);
                }
                if (Bills != null && Bills.Length > 0)
                {
                    if (Bills[0].Check(cost))
                    {
                        Bills[0].Add(cost);
                    }
                    else
                    {
                        Console.WriteLine("Рахунок перевищено. Неможливо додати більше до рахунку.");
                    }
                }
            }
        }

        public void Message(int quantity, Customer other)
        {
            if (Operators != null && Operators.Length > 0 && other != null && other.Operators != null)
            {
                if (Operators[0] == other.Operators[0])
                {
                    double cost = Operators[0].CalculateMessageCost(quantity, this, other);
                    if (Bills != null && Bills.Length > 0)
                    {
                        if (Bills[0].Check(cost))
                        {
                            Bills[0].Add(cost);
                        }
                        else
                        {
                            Console.WriteLine("Рахунок перевищено. Неможливо додати більше до рахунку.");
                        }
                    }
                }
            }
        }
        public void Connection(double amount)
        {
            if (Operators != null && Operators.Length > 0)
            {
                double cost = Operators[0].CalculateNetworkCost(amount);
                if (Bills != null && Bills.Length > 0)
                {
                    if (Bills[0].Check(cost))
                    {
                        Bills[0].Add(cost);
                    }
                    else
                    {
                        Console.WriteLine("Рахунок перевищено. Неможливо додати більше до рахунку.");
                    }
                }
            }
        }
        public void PayBill(double amount)
        {
            if (Bills != null && Bills.Length > 0)
            {
                if (amount <= Bills[0].CurrentDebt)
                {
                    Bills[0].Pay(amount);
                }
                else
                {
                    Console.WriteLine("Сума платежу перевищує поточний борг.");
                }
            }
            else
            {
                Console.WriteLine("Неможливо здійснити платіж. Немає рахунку, пов'язаного з клієнтом.");
            }
        }
        public void ChangeOperator(Operator newOperator)
        {
            if (Operators != null && Operators.Length > 0 && newOperator != null)
            {
                Operators[0] = newOperator;
            }
        }

        public void ChangeBillLimit(double newLimit)
        {
            if (Bills != null && Bills.Length > 0)
            {
                Bills[0].ChangeTheLimit(newLimit);
            }
        }
        public double GetUsedAmount()
        {
            if (Bills != null && Bills.Length > 0)
            {
                return Bills[0].CurrentDebt;
            }
            return 0.0;
        }
        public double GetRemainingAmount()
        {
            if (Bills != null && Bills.Length > 0)
            {
                return LimitingAmount - Bills[0].CurrentDebt;
            }
            return LimitingAmount;
        }
    }
}

File: Пащук Нікіта/Lab1/Lab_1/Customer.py
from Operator import Operator
from Bill import Bill


class Customer:
    def __init__(self, id: int, name: str, age: int, operator: Operator, bill: Bill):
        self.id = id
        self.name = name
        self.age = age
        self.Operator = operator
        self.Bill = bill

    def talk(self, minute: int, other_customer: 'Customer'):
        if self.Bill.check(self.Bill.current_debt):
            self.Bill.add(self.Operator.calculate_talking_cost(minute, self))
            return f"Customer {self.name} talked to customer {other_customer.name}"
        else:
            return "Limited amount was exceeded, operation was not done"

    def message(self, quantity: int, other_customer: 'Customer'):
        if self.Bill.check(self.Bill.current_debt):
            self.Bill.add(self.Operator.calculate_message_cost(quantity, self, other_customer))
            return f"Customer {self.name} sent message to customer {other_customer.name}"
        else:
            return "Limited amount was exceeded, operation was not done"

    def connection(self, amount: float):
        if self.Bill.check(self.Bill.current_debt):
            self.Bill.add(self.Operator.calculate_network_cost(amount))
            return f"Customer has connected to the internet while having {amount} MB"
        else:
            return "Limited amount was exceeded, operation was not done"

    def change_operator(self, Operators, wanted_id):
        current_id = self.Operator.id
        self.Operator = Operators[wanted_id]
        return f"Operator was changed from {current_id} to {wanted_id}"
